gnucash/gnome-utils/gnome-utils.scm

```diff
@@ -41,3 +41,23 @@
 
 (load-from-path "gnc-menu-extensions")
 
+(define-public gnc:set-ui-status
+  (let ((save-warn gnc:gui-warn)
+        (save-error gnc:gui-error)
+        (save-msg gnc:gui-msg))
+    (lambda (status)
```

Comment:
I'm a bit puzzled how "status" gets set. Can you explain this a bit ?

I see you're calling scm_call_1 ... in in gnc_ui_start_event_loop with an SCM boolean, but I can't understand how this eventually ends up in the "status" variable here.

Reply:
It's a lispy closure.

1. Two ways of defining a function in guile. `(define (f x) ...)` or `(define f (lambda (x) ...))` These two are *exactly* the same. `x` is hopefully useful in `f`.

2. Any scheme expression can create a local environment using `(let ((vars..)) ...)` - I use it extensively to create intermediate values instead of long complicated expressions; vars are only visible within the let environment, and *discarded* when the (let) is finished.

Combining 1 and 2 leads to
```
(define gui-set
  (let ((old-fn gnc:gui-warn))
    (lambda (status)
      (if status (set! gnc:gui-warn ...)
                 (set! gnc:gui-warn old-fn))
      ...)))
```
whereby gui-set is a *function* (see the lambda) being created within the local environment containing old-fn which is accessible only *within* the `(let)` environment. So gui-set is now a lambda, accepting 1 argument `status` which receives #t/#f, and uses it to modify *global* gnc:gui-warn and can also save it in *local* old-fn.

Very schemey and compact. Avoids using global variables. No tricks, no pythonic generators. Similar to https://www.gnu.org/software/guile/manual/html_node/Shared-Variable.html#Shared-Variable

Reply:
Very schemey and compact... and very hard to grasp for a non schemer :(
I'll try to remember next time I see such a construct...

Reply:
There's lots of previous examples in code using a more elaborate closure modelled from the manual
https://www.gnu.org/software/guile/manual/html_node/OO-Closure.html#OO-Closure

e.g. https://github.com/Gnucash/gnucash/blob/3a105f0728984df7f063110acc8390c93722d581/gnucash/report/report-system/report-utilities.scm#L293 whereby `commoditylist` is the local variable

https://github.com/Gnucash/gnucash/blob/3a105f0728984df7f063110acc8390c93722d581/libgnucash/app-utils/options.scm#L204 whereby `value` and `value->string` are local variables... and are miles uglier than the compact stuff above...